lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/avro/ProtocolSerializer.java

```diff
@@ -0,0 +1,196 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.reef.wake.avro;
+
+import org.apache.avro.io.*;
+import org.apache.avro.specific.SpecificDatumReader;
+import org.apache.avro.specific.SpecificDatumWriter;
+import org.apache.avro.specific.SpecificRecord;
+import org.apache.avro.specific.SpecificRecordBase;
+import org.apache.reef.wake.MultiObserver;
+import org.apache.reef.wake.avro.message.Header;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
+import io.github.lukehutch.fastclasspathscanner.scanner.ScanResult;
+
+/**
+ * The ProtocolSerializer generates serializers and deserializers for
+ * all of the Avro messages contained in a specified package.
+ */
+public final class ProtocolSerializer {
+  private static final Logger LOG = Logger.getLogger(ProtocolSerializer.class.getName());
+  // Maps for mapping message class names to serializer and deserializer classes.
+  private Map<String, MessageSerializer> nameToSerializerMap;
+  private Map<String, MessageDeserializer> nameToDeserializerMap;
+
+  public ProtocolSerializer() {
+    nameToSerializerMap = new HashMap<>();
+    nameToDeserializerMap = new HashMap<>();
+  }
+
+  /**
+   * Finds all of the messages in the specified packaged and calls register.
+   * @param messagePackage A string which contains the full name of the
+   *                       package containing the protocol messages.
+   */
+  public void initialize(final String messagePackage) {
+    LOG.log(Level.INFO, "Start: Serializer.Initialize");
+
+    // Build a list of the message reflection classes.
+    final ScanResult scanResult = new FastClasspathScanner(messagePackage).scan();
+    final List<String> scanNames = scanResult.getNamesOfSubclassesOf(SpecificRecordBase.class);
+    final List<Class<?>> messageClasses = scanResult.classNamesToClassRefs(scanNames);
+    LOG.log(Level.INFO, "Number of messages = " + Integer.toString(messageClasses.size()));
+
+    try {
+      // Register all of the messages in the specified package.
+      for (Class<?> cls : messageClasses) {
+        LOG.log(Level.INFO, "Found message class: " + cls.getName() + " " + cls.getSimpleName());
+        Method register = ProtocolSerializer.class.getMethod("register", cls.getClass());
+        LOG.log(Level.INFO, "Obtained the method class instance");
+        register.invoke(this, cls);
+      }
+      // Register the header message.
+      Method register = ProtocolSerializer.class.getMethod("register", Header.class);
```

Comment:
I removed these two lines and added.

messageClasses.add(Header.class);

before the loop.